libp2p/transports/wstransport.nim

```diff
@@ -0,0 +1,199 @@
+## Nim-LibP2P
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [Defect].}
+
+import std/[sequtils]
+import chronos, chronicles
+import transport,
+       ../errors,
+       ../wire,
+       ../multicodec,
+       ../multistream,
+       ../connmanager,
+       ../multiaddress,
+       ../stream/connection,
+       ../upgrademngrs/upgrade,
+       websock/websock
+
+logScope:
+  topics = "libp2p wstransport"
+
+export transport
+
+const
+  WsTransportTrackerName* = "libp2p.wstransport"
+
+type
+  WsStream = ref object of Connection
+    session: WSSession
+
+proc init*(T: type WsStream,
+           session: WSSession,
+           dir: Direction,
+           timeout = 10.minutes,
+           observedAddr: MultiAddress = MultiAddress()): T =
+
+  let stream = T(session: session,
+             timeout: timeout,
+             dir: dir,
+             observedAddr: observedAddr)
+  stream.initStream()
+  return stream
+
+method readOnce*(
+  s: WsStream,
+  pbytes: pointer,
+  nbytes: int):
+  Future[int] =
+  return s.session.recv(pbytes, nbytes)
+
+method write*(s: WsStream, msg: seq[byte]):
+  Future[void] {.async.} =
+  #TODO the write procs can raise LPStreamEOF
```

Comment:
Unfortunately, chronos don't yet support specific annotations on async transformations. See https://github.com/status-im/nim-libp2p/pull/547, most of what's needed to support it is already there.